Refactor/fix bugs rules

danssaul opened the pull request:
# Fix Bugs rules

## Description

🐛 Resumo dos bugs encontrados:
Ao criar um usuário ou uma academia (gym) com um nome já existente em um registro inativo, o sistema retorna 409 Conflict – deveria permitir, já que está inativo.

Test if it’s possible to create a gym with a inactive user

Na rota PATCH, ao tentar alterar o nome para o nome de um usuário ou gym inativo, também recebemos 409 Conflict – o nome deveria estar disponível.

A rota PATCH /users/{id} permite atualizar o nome de um usuário que está inativo, o que não deveria acontecer.


## Checklist
- [x] All acceptance criteria have been met
- [x] My code follows the project's coding style.
- [ ] I have documented my code
- [x] I wrote unit tests for the changes that I made

## Task in Notion
[task on notion](https://www.notion.so/16bb8830dc7b8068807af79065c08418?v=16bb8830dc7b801292da000c63c754b1&p=1dcb8830dc7b80c18c83d1d5aa338eae&pm=s)

danssaul committed:
Refactor: Update gym existence check to include status in GymService and tests

danssaul committed:
Refactor: Update user existence checks to include status in UserRepository and UserService

danssaul committed:
test: Add test for updating inactive user to throw UserNotFoundException

danssaul committed:
Refactor: Remove redundant logging for duplicate gym and user creation checks

danssaul committed:
Refactor: Remove warning log for duplicate user creation check

tomerpeceniski reviewed (commented):
Hey Daniel, every thing ok. The unique bug that was not fixed is:
-> When trying to update gym name, is not possible to use a gym name that is being used by an inactivated gym.

Can you fix it before approving ?

danssaul committed:
Merge branch 'main' into refactor/fiz-bugs-rules

danssaul committed:
refactor: Update gym existence check to include status in GymRepository and GymService

danssaul committed:
Merge branch 'main' into refactor/fiz-bugs-rules

danssaul committed:
feat: Enhance GymService to manage workout inactivation upon gym deletion

danssaul committed:
test: Add workout repository mock to GymServiceTest for gym deletion verification

danssaul committed:
test: Add unit test for updating inactive gym in GymControllerTest

danssaul committed:
test: Update GymServiceTest to verify gym deletion inactivates associated workouts and exercises

itaayp commented on src/main/java/igym/services/GymService.java:
```diff
@@ -78,15 +82,37 @@ public void deleteGym(UUID id) {
             throw new GymNotFoundException("Gym with id " + id + " not found");
         }
 
+        inactivateWorkouts(gym);
```
👀

itaayp replied:
It seems that this change should be in a different PR 😄

itaayp commented on src/main/java/igym/services/GymService.java:
```diff
@@ -78,15 +82,37 @@ public void deleteGym(UUID id) {
 
+        inactivateWorkouts(gym);
         gym.setStatus(Status.inactive);
         gymRepository.save(gym);
```
Because you are doing many transactions in DB, we need to implement the `@transactional` Spring anotation

itaayp commented on src/main/java/igym/services/GymService.java:
```diff
@@ -78,15 +82,37 @@ public void deleteGym(UUID id) {
+        workouts.forEach(workout -> {
+            if (workout.getStatus() == Status.active) {
+                workout.setStatus(Status.inactive);
+                if (workout.getExerciseList() != null) {
+                    workout.getExerciseList().forEach(ex -> {
+                        ex.setStatus(Status.inactive);
+                        logger.debug("Exercise {} inactivated", ex.getId());
+                    });
+                }
```
To avoid inactivating an already inactivated exercise, we need to check this condition before doing the change. Something similar to the 101.

And also, to avoid nested conditionals, I recomend to create a new method called `inactivateExercises`.

itaayp commented on src/main/java/igym/services/GymService.java:
```diff
@@ -78,15 +82,37 @@ public void deleteGym(UUID id) {
+            }
+        });
+
+        workoutRepository.saveAll(workouts); // Save all updated workouts
```
```suggestion
        workoutRepository.saveAll(workouts);
        logger.info("Workouts deleted")
```

itaayp commented on src/main/java/igym/services/GymService.java:
```diff
@@ -48,17 +51,18 @@ public Gym updateGym(UUID id, String name) {
             throw new GymNotFoundException("Gym with id " + id + " not found");
         }
 
+        if (gymRepository.existsByNameAndStatus(name, Status.active)) {
```
Why for creating Gyms, we check name, id and status, but for updating we only check name and status?

itaayp commented on src/main/java/igym/repositories/GymRepository.java:
```diff
@@ -5,11 +5,11 @@
-    boolean existsByName(String name);
-
-    boolean existsByNameAndUserId(String name, UUID userId);
+    boolean existsByNameAndStatus(String name, Status status);
+    boolean existsByNameAndUserIdAndStatus(String name, UUID userId, Status status);
```
Daniel, as we are always checking for the "active" status, I suggest removing the status parameter but adding this check inside the method.

tomerpeceniski replied:
Itay, I think it makes sense to maintain this Status check here. Because we are looking for other gyms that have this name and have status active.

itaayp commented on src/main/java/igym/services/UserService.java:
```diff
@@ -74,7 +74,13 @@ public User updateUser(UUID id, String name) {
-        if (repository.existsByName(name)) {
+
+        if (user.getStatus() == Status.inactive) {
+            logger.warn("User with id {} is inactive", id);
```
```suggestion
            logger.info("User with id {} is inactive", id);
```

itaayp commented on src/test/java/igym/services/GymServiceTest.java:
I suggest to test the folowing scenarios:

**Scenario 1**
1. Gym named "A" is inactive
2. Try to create a gym named "A"

**Scenario 2**
1. Gym named "A" is inactive
2. Gym named "B" is active
3. Try to change the gym name "B" to "A"

tomerpeceniski replied:
But we cannot do this kind of test because we are mocking our repository. We define our responses from our repository. Which means that when we run the method:

repository.existsByNameAndUserIdAndStatus() we define that it will return false, because there is no gym with this name, this status and this user. It is not really looking for those gyms in the DB. So creating gyms in the repository does not make difference.

itaayp commented on src/test/java/igym/services/UserServiceTest.java:
I suggest to test the folowing scenarios:

**Scenario 1**

1. User named "A" is inactive
2. Try to create a User named "A"

**Scenario 2**
1. User named "A" is inactive
2. User named "B" is active
3. Try to change the User name "B" to "A"

tomerpeceniski replied:
The same comment that I made in the GymServiceTest is valid here

tomerpeceniski commented on src/main/java/igym/services/GymService.java:
```diff
@@ -81,11 +85,34 @@ public void deleteGym(UUID id) {
     }
-    
+
+    private void inactivateWorkouts(Gym gym) {
+        List<Workout> workouts = workoutRepository.findByGym(gym);
+        if (workouts.stream().anyMatch(w -> w.getStatus() == Status.active)) {
+            logger.info("Inactivating {} workouts for gym {}", workouts.size(), gym.getId());
+        }
+
+        workouts.forEach(workout -> {
+            if (workout.getStatus() == Status.active) {
+                workout.setStatus(Status.inactive);
+                if (workout.getExerciseList() != null) {
+                    workout.getExerciseList().forEach(ex -> {
+                        ex.setStatus(Status.inactive);
+                        logger.debug("Exercise {} inactivated", ex.getId());
+                    });
+                }
```
I suggest, instead of creating this method here, use the method deleteWorkout() from WorkoutService class. So then we don't need to create this logic of deletion again. Also, in the deleteWorkout() there are already the logs and business rules. And we guarantee that the WorkoutService will be responsible for managing workouts.

tomerpeceniski committed:
Fix: Update updateGym method to check gym names for the same user

tomerpeceniski committed:
Refactor: update GymService and WorkoutService to improve workout inactivation logic and update tests

- Send workout inactiovation logic to WorkoutService
- Refactor WorkoutService to avoid mutual depencence of GymService
- Refactor tests to match new architecture

tomerpeceniski committed:
Feat: Add logic of deleting workout and exercises when deleting user

- Pass list of gym deletion logic from UserService to GymService
- Change GymService structure for avoiding mutual dependence
- Refactor tests for new structure

tomerpeceniski commented:
@itaayp @daniel1sender I've made some updated after the reviews:

- Added check by user when validating if there is already a gym with the new name (GymService.updateGym)
- Pass logic of deleting list of workouts from GymService to WorkoutService
- Pass logic of deleting list of gyms from UserService to GymService (now, when deleting a user we are deleting also its gyms, workouts and exercises)
- Refactor services for avoiding mutual dependences (it breaks Spring when there is a mutual dependence)
- Refactor unit tests for the new structure

danssaul committed:
chore: Update .gitignore to include node_modules and package-lock.json

danssaul commented on src/main/java/igym/services/GymService.java:
```diff
@@ -46,19 +52,15 @@ public Gym createGym(Gym gym, UUID userId) {
-            throw new GymNotFoundException("Gym with id " + id + " not found");
+
+        if (gymRepository.existsByNameAndUserIdAndStatus(name, gym.getUser().getId() ,Status.active)) {
+            throw new DuplicateGymException("A gym with the name '" + name + "' already exists for this user");
```
I thought one user could have more than one gym with the same name, but it is better to avoid this previous definition if it generates inconsistent behavior. For now I don't see problem in both of the ways because we are searching the for the id first.

danssaul commented on src/main/java/igym/services/WorkoutService.java:
```diff
@@ -68,28 +71,51 @@ public List<Workout> getWorkoutsByGymId(UUID gymId) {
+        logger.info("Workout with id {} have been inactivated", workoutId);
+    }
+
+    private void deleteExerciesList(Workout workout) {
```
why different names for this method but the same propose?

danssaul commented on src/main/java/igym/services/GymService.java:
```diff
@@ -73,20 +75,25 @@ public List<Gym> findAllGyms() {
     }
-    
+
+    private void inactivateWorkouts(Gym gym) {
+        List<Workout> workouts = workoutRepository.findByGym(gym);
+        if(workouts != null) {
+            workouts.forEach(w -> workoutService.deleteWorkout(w.getId()));
+        }
+        workoutRepository.saveAll(workouts);
+    }
+
```
there is an issue here, if the workout list is empty I don't think this gonna work, the shape of this method is a bit weird to me

danssaul commented on src/main/java/igym/services/GymService.java:
```diff
@@ -73,20 +75,25 @@ public List<Gym> findAllGyms() {
+    private void inactivateWorkouts(Gym gym) {
+        List<Workout> workouts = workoutRepository.findByGym(gym);
+        if(workouts != null) {
+            workouts.forEach(w -> workoutService.deleteWorkout(w.getId()));
```
This method is called deleting the gym and deleting the workout by ID. But if there are no workouts or workouts are inactive I cannot delete the gym.

tomerpeceniski committed:
Refactor: Remove unused WorkoutRepository and update gym and workout deletion logic

tomerpeceniski committed:
Merge branch 'refactor/fiz-bugs-rules' of https://github.com/tomerpeceniski/igym into refactor/fiz-bugs-rules

tomerpeceniski committed:
Merge branch 'main' into refactor/fiz-bugs-rules

tomerpeceniski committed:
refactor: improve and add unit tests for better coverage
